### PR TITLE
Models/coral_corn

### PR DESCRIPTION
**Pull request by tunczyk101**

### CODE REVIEW

File: dataset/creating_dataset.py [hunk 1]
```diff
@@ -182,6 +183,18 @@ def extract_and_assign_chars(
         df[char] = bestiary[path_to_char].apply(get_value)
 
 
+def get_max_spell_level(items_list: list[dict]):
```
Comment (j-adamczyk) [inline dataset/creating_dataset.py:186]: `-> int`

File: training/models/coral_corn.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        probas = torch.sigmoid(logits)
+
+        return logits, probas
+
+    def predict_proba(self, x):
+        return sigmoid(self(x))
+
+    def predict(self, x, threshold: float = 0.5):
+        y_pred_score = self.predict_proba(x)
+        return (y_pred_score > threshold).to(torch.int32)
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:60]: >=

File: training/models/coral_corn.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+class Coral(BaseEstimator, ClassifierMixin):
+    def __init__(
+        self,
+        input_size=53,
+        num_classes=NUM_CLASSES,
+        lambda_reg=0.01,
+        learning_rate=0.05,
+        num_epochs=100,
+        batch_size=128,
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:71]: Add types

File: training/models/coral_corn.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        self.num_epochs = num_epochs
+        self.batch_size = batch_size
+        self.num_classes = num_classes
+        self.model = CORAL_MLP(input_size=self.input_size, num_classes=num_classes)
+        self.model.to(DEVICE)
+        self.optimizer = torch.optim.AdamW(
+            self.model.parameters(), lr=learning_rate, weight_decay=self.lambda_reg
+        )
+
+    def fit(self, X, y):
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:86]: This would probably be quite easier with `skorch`

File: training/models/coral_corn.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    def __init__(self, input_size: int, num_classes: int):
+        super().__init__()
+        self.network = nn.Sequential(
+            nn.Linear(input_size, 100),
+            nn.ReLU(),
+            nn.Linear(100, 50),
+            nn.ReLU(),
+            ### Specify CORN layer
+            torch.nn.Linear(50, (num_classes - 1)),
+        )
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:140]: Like above, maybe 128 and 64?

File: training/models/coral_corn.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+    def forward(self, x):
+        logits = self.network(x)
+
+        return logits
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:145]: `return self.network(x)`

File: training/models/coral_corn.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+class Corn(BaseEstimator, ClassifierMixin):
+    def __init__(
+        self,
+        input_size=53,
+        num_classes=NUM_CLASSES,
+        lambda_reg=0.01,
+        learning_rate=0.05,
+        num_epochs=100,
+        batch_size=128,
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:156]: Types

File: training/models/coral_corn.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    ):
+        torch.manual_seed(RANDOM_STATE)
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:158]: Seed in `.fit()`, not in constructor

File: training/models/coral_corn.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:73]: Seed in `.fit()`, not in constructor

File: training/models/coral_corn.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        )
+
+    def fit(self, X, y):
+        train_dataset = OrdinalDataset(X.to_numpy(), y.to_numpy())
+        train_loader = DataLoader(
+            dataset=train_dataset,
+            batch_size=self.batch_size,
+            shuffle=True,  # want to shuffle the dataset
+            num_workers=0,
+        )  # number processes/CPUs to use
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:178]: One line above?

File: training/models/coral_corn.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        for epoch in range(self.num_epochs):
+
+            self.model = self.model.train()
+            for batch_idx, (features, class_labels) in enumerate(train_loader):
+
+                class_labels = class_labels.to(DEVICE)
+                features = features.to(DEVICE)
+                logits = self.model(features)
+
+                #### CORN loss
```
Comment (j-adamczyk) [inline training/models/coral_corn.py:189]: Why comment?

**Review by j-adamczyk — CHANGES_REQUESTED** [review]

File: training/models/coral_corn.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+            shuffle=True,  # want to shuffle the dataset
+            num_workers=0,
+        )  # number processes/CPUs to use
+
+        for epoch in range(self.num_epochs):
+
+            self.model = self.model.train()
+            for batch_idx, (features, class_labels) in enumerate(train_loader):
+
+                levels = levels_from_labelbatch(
```
[the remark this answers is not in the document]
Comment (tunczyk101) [reply training/models/coral_corn.py:100]: Contact to author:
# coral_pytorch
# Author: Sebastian Raschka <sebastianraschka.com>

**Comment by tunczyk101** [comment]
Obecnie tylko Coral
Ten skorch z jakiegoś powodu jest dla mnie większą drogą przez mękę niż wszystko do tej pory.
Błagam o potwierdzenie, że jest ok lub o informacje co jest nie tak i dodam Corna

File: training/models/coral_corn.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+from coral_pytorch.layers import CoralLayer
+from coral_pytorch.losses import coral_loss, corn_loss
+from sklearn.base import BaseEstimator, ClassifierMixin
+from torch import nn, sigmoid
+from torch.utils.data import DataLoader, Dataset
+
+from training.constants import RANDOM_STATE
+
+
+NUM_CLASSES = 53
```
[the remark this answers is not in the document]
Comment (j-adamczyk) [reply training/models/coral_corn.py:21]: +1

**Review by j-adamczyk — APPROVED** [review]
